src/layers/domain/repository/device_repository/v2.py

```diff
@@ -0,0 +1,190 @@
+# from collections import defaultdict
+from functools import partial
+
+from attr import asdict as _asdict
+from domain.core.device.v2 import (
+    Device,
+    DeviceCreatedEvent,
+    DeviceKeyAddedEvent,
+    DeviceKeyDeletedEvent,
+    DeviceUpdatedEvent,
+)
+from domain.core.device_key.v2 import DeviceKey
+from domain.repository.errors import ItemNotFound
+from domain.repository.keys import TableKeys
+from domain.repository.marshall import marshall, marshall_value, unmarshall
+from domain.repository.repository.v2 import Repository
+from domain.repository.transaction import (
+    ConditionExpression,
+    TransactionStatement,
+    TransactItem,
+)
+
+
+def asdict(obj) -> dict:
+    return _asdict(obj, recurse=False)
+
+
+def _dynamodb_update_expression(updated_fields: dict):
+    expression_attribute_names = {}
+    expression_attribute_values = {}
+    update_clauses = []
+
+    for field_name, value in updated_fields.items():
+        field_name_placeholder = f"#{field_name}"
+        field_value_placeholder = f":{field_name}"
+
+        update_clauses.append(f"{field_name_placeholder} = {field_value_placeholder}")
+        expression_attribute_names[field_name_placeholder] = field_name
+        expression_attribute_values[field_value_placeholder] = marshall_value(value)
+
+    update_expression = "SET " + ", ".join(update_clauses)
+
+    return dict(
+        UpdateExpression=update_expression,
+        ExpressionAttributeNames=expression_attribute_names,
+        ExpressionAttributeValues=expression_attribute_values,
+    )
+
+
+def _device_primary_keys(device_id: str, device_keys: list[DeviceKey]) -> list[dict]:
+    """
+    Generates all the fully marshalled (i.e. {"pk": {"S": "123"}} DynamoDB
+    primary keys (i.e. pk + sk) for the provided Device. This is one primary key
+    plus an additional primary key for every value of Device.keys
+    """
+    root_primary_key = marshall(
+        pk=TableKeys.DEVICE.key(device_id), sk=TableKeys.DEVICE.key(device_id)
+    )
+    non_root_primary_keys = [
+        marshall(pk=pk, sk=pk)
+        for pk in (TableKeys.DEVICE.key(k.key_type, k.key_value) for k in device_keys)
+    ]
+    return [root_primary_key] + non_root_primary_keys
+
+
+def handle_update_device(
+    table_name: str, id: str, keys: list[DeviceKey], data: dict
+) -> list[TransactItem]:
+    primary_keys = _device_primary_keys(device_id=id, device_keys=keys)
+    update_expression = _dynamodb_update_expression(updated_fields=data)
+    update_statement = partial(
```

Comment:
This is for my own understanding - how much value does having a partial here achieve? feels like if the update_statement was being used more than once then i can understand its use. I might be missing what this actually achieves though :D